### PR TITLE
fix(gatsby-plugin-sharp): prevent crashing on invalid cropFocus

### LINKED ISSUE #32720: Setting cropFocus to attention or entropy crashes Gatsby
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

You can set custom cropFocus strategies with gatsby-plugin-image (see https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image/#transformoptions)

This generally works fine but if you try to set it to `entropy` or `attention` it fails with this error:

```
error converting image Processing
/Users/kylemathews/programs/gatsby/examples/using-gatsby-image/.cache/caches/gatsby-plugin-image/db148be6b48f9dfcfa151d035fc34218/photo-1597305877032-0668b3c6413a.jpg failed

Original error:

Expected valid fit for fit but received top of type string



  WorkerError: Processing /Users/kylemathews/programs/gatsby/examples/using-gatsby-image/.cache/caches/gatsby-plugin-image/db148be6b48f9dfcfa151d035fc34218/photo-1597305877032-0668b3c6413a.jpg fai
  led
  Original error:
  Expected valid fit for fit but received top of type string

  - manager.ts:291 enqueueJob
    [using-gatsby-image]/[gatsby]/src/utils/jobs/manager.ts:291:13
```

The problem is attention/entropy are numeric codes defined in the sharp package. `attention` is the default and we set it like this:

https://github.com/gatsbyjs/gatsby/blob/59a6287a11eb5b136e3e681e7d1d1c08d169b601/packages/gatsby-plugin-sharp/src/image-data.ts#L117

So for user set `attention`/`entropy` settings to work, we need to check for those strings and then replace it with the var set on the `sharp` package.

### Reproduction Link

https://github.com/gatsbyjs/gatsby/blob/master/examples/using-gatsby-image/README.md

### Steps to Reproduce

Create a `<StaticImage>` component with `cropFocus` set to `entropy` like:

```
<StaticImage
  src="image.jpg"
  transformOptions={{cropFocus: `entropy`}}
/>
```

Notice that `cropFocus: "north"` works (as do other strategies https://sharp.pixelplumbing.com/api-resize#resize

### Expected Result

No error 

### Actual Result

Error

### Environment

```shell
n/a
```

### Config Flags

_No response_

### PR DESCRIPTION
**Pull request by nategiraudeau**
## Description

Checks for invalid `cropFocus` value and reports error instead of crashing app:

```jsx
<StaticImage
    src="image.png"
    transformOptions={{ cropFocus: `INVALID_CROPFOCUS` }} />
```

It now displays the following error instead of causing the entire app to crash:

```
ERROR 

Invalid cropFocus option 'INVALID_CROPFOCUS' specified for image.png.
See https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image/#transformoptions
for more information about cropFocus.
```

## Related Issues

Closes #32720

### CODE REVIEW

File: packages/gatsby-plugin-sharp/src/image-data.ts [hunk 1]
```diff
@@ -115,11 +115,40 @@ export async function generateImageData({
...
+    `southeast`,
+    `south`,
+    `southwest`,
+    `west`,
+    `northwest`,
+    `center`,
+    `centre`,
+    `entropy`,
+    `attention`,
+  ]
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/image-data.ts:143]: Can we change it into the following? This makes the link more dynamic as it gets it values from sharp so we don't have to maintain it. Could you also add a unit test to test this behavior?
 
```suggestion
  const cropFocusOptions = Object.keys(sharp.position).concat(Object.keys(sharp.gravity)).concat(Object.keys(sharp.strategy))
```

File: packages/gatsby-plugin-sharp/src/image-data.ts [hunk 2]
```diff
@@ -115,11 +115,40 @@ export async function generateImageData({
...
+
+  if (typeof cropFocus !== `number` && !cropFocusOptions.includes(cropFocus)) {
+    reporter.error(
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/image-data.ts:146]: What was previous behavior? Did it also fail or did it go through? I feel like we should fail during gatsby build but not during develop. What do you think?
```suggestion
    reporter.panicOnBuild(
```

**Review by wardpeet — COMMENTED** [review]
Nice one! Thank you for you contribution, I've added a review.

**Comment by nategiraudeau** [comment]
Thanks

File: packages/gatsby-plugin-sharp/src/image-data.ts (hunk 1 above)
Comment (nategiraudeau) [reply packages/gatsby-plugin-sharp/src/image-data.ts]: Thanks. `sharp.position` doesn't actually exist, but `sharp.gravity` and `sharp.strategy` do, so I had to keep the hardcoded values for position:
```ts
const cropFocusOptions = [
  `top`,
  `right top`,
  `right`,
  `right bottom`,
  `bottom`,
  `left bottom`,
  `left`,
  `left top`,
]
  .concat(Object.keys(sharp.gravity))
  .concat(Object.keys(sharp.strategy))
```

Comment (nategiraudeau) [reply packages/gatsby-plugin-sharp/src/image-data.ts]: I can also make a unit test.

File: packages/gatsby-plugin-sharp/src/image-data.ts (hunk 2 above)
Comment (nategiraudeau) [reply packages/gatsby-plugin-sharp/src/image-data.ts]: Previously, the app would crash. Also failing during build and not during develop sounds like a good idea.

File: packages/gatsby-plugin-sharp/src/image-data.ts (hunk 1 above)
Comment (wardpeet) [reply packages/gatsby-plugin-sharp/src/image-data.ts]: Interesting cause for me this works:
```
$ node -e "console.log(Object.keys(require('sharp').position))"
[
  'top',
  'right',
  'bottom',
  'left',
  'right top',
  'right bottom',
  'left bottom',
  'left top'
]
```

Unit tests are always great 😍 But feel free to do it in a follow up

**Comment by wardpeet** [comment]
You can ignore integration_tests_gatsby_source_wordpress  for now.

> ci/circleci: integration_tests_images 
Seems to have snapshot issues, if it's helpful I can update them for you.

File: packages/gatsby-plugin-sharp/src/image-data.ts (hunk 1 above)
Comment (nategiraudeau) [reply packages/gatsby-plugin-sharp/src/image-data.ts]: Screen Shot 2021-09-13 at 4 42 15 PM

I think it's just a typescript problem that can just be ignored. I'll change it back to `Object.keys(sharp.position)`. Also I'll try to add the unit test in another pull request sometime soon.

**Review by wardpeet — APPROVED** [review]

**Comment by nategiraudeau** [comment]
@wardpeet is there anything else I need to do to get this merged?

**Comment by wardpeet** [comment]
No, I'll fix the ci/circleci: integration_tests_images  tests

**Comment by LekoArts** [comment]
Thanks for the PR! This was fixed in the meantime: https://github.com/gatsbyjs/gatsby/issues/32720#issuecomment-1330632191

**Pull request closed without merging**